extra_data/components.py

```diff
@@ -926,6 +738,311 @@ def zip_trains_pulses(trains, pulses):
+
+    @property
+    def ndim(self):
+        return self._eg_keydata.ndim + 1
+
+    @property
+    def shape(self):
+        return ((len(self.modno_to_keydata), len(self.train_id_coordinates()))
+                + self._eg_keydata.entry_shape)
+
+    @property
+    def dimensions(self):
+        return ['module', 'trainId'] + ['dim_%d' % i for i in range(self.ndim - 2)]
+
+    @property
+    def dtype(self):
+        return self._eg_keydata.dtype
+
+    def _with_selected_det(self, det_selected):
+        # Overridden for XtdfImageMultimodKeyData to preserve pulse selection
+        return MultimodKeyData(det_selected, self.key)
+
+    def select_trains(self, trains):
+        return self._with_selected_det(self.det.select_trains(trains))
+
+    def split_trains(self, parts=None, trains_per_part=None, frames_per_part=None):
+        for det_split in self.det.split_trains(parts, trains_per_part, frames_per_part):
+            yield self._with_selected_det(det_split)
+
+    def ndarray(self, *, fill_value=None, out=None, roi=(), astype=None, module_gaps=False):
+        """Get data as a plain NumPy array with no labels"""
+        train_ids = np.asarray(self.det.train_ids)
+
+        module_dim = self.det.n_modules if module_gaps else len(self.modno_to_keydata)
+
+        out_shape = ((module_dim, len(train_ids))
+                     # Shape of 1 frame for 1 module with the ROI applied:
+                     + roi_shape(self._eg_keydata.entry_shape, roi))
+
+        if out is None:
+            dtype = self._eg_keydata.dtype if astype is None else np.dtype(astype)
+            out = _out_array(out_shape, dtype, fill_value=fill_value)
+        elif out.shape != out_shape:
+            raise ValueError(f'requires output array of shape {out_shape}')
+
+        for i, (modno, kd) in enumerate(sorted(self.modno_to_keydata.items())):
+            mod_ix = (modno - self.det._modnos_start_at) if module_gaps else i
+            for chunk in kd._data_chunks:
+                for tgt_slice, chunk_slice in self.det._split_align_chunk(chunk, train_ids):
+                    chunk.dataset.read_direct(
+                        out[mod_ix, tgt_slice], source_sel=(chunk_slice,) + roi
+                    )
+        return out
+
+    def xarray(self, *, fill_value=None, roi=(), astype=None):
+        from xarray import DataArray
+        arr = self.ndarray(fill_value=fill_value, roi=roi, astype=astype)
+
+        coords = {'module': self.modules, 'trainId': self.train_id_coordinates()}
+        return DataArray(arr, dims=self.dimensions, coords=coords)
+
+    def dask_array(self, *, labelled=False, fill_value=None, astype=None):
+        from dask.delayed import delayed
+        from dask.array import concatenate, from_delayed
+
+        entry_size = (self.dtype.itemsize *
+            len(self.modno_to_keydata) * np.product(self._eg_keydata.entry_shape)
+        )
+        # Aim for 1GB chunks, with an arbitrary maximum of 256 trains
```

Comment:
Why 1 GiB? Any sense in making it configurable?

Reply:
It's entirely guesswork. I think Dask might prefer smaller chunks, but then you're also more likely to end up opening the same file on several nodes, because it's been split up into many tiny chunks.

I've got a separate PR #322 to let Dask pick the chunk size for `KeyData.dask_array()`, which would also make it configurable through a Dask config file. But its default is 128 MiB, and I guess that spreading that over 16 modules and loading 8 MiB at a time might be smaller than ideal.

Reply:
Thanks.

It's not entirely scientific, but backed by anecdotal evidence and expert's suspicions: dCache does not like parallel access to the same file, but nicely scales with different nodes accessing different files. So it might be interesting to pick the default based on the storage location.